Aplib.Core/Belief/MemoryBelief.cs

```diff
@@ -0,0 +1,104 @@
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="MemoryBelief{TReference, TResource}"/> class.
+        /// </summary>
+        /// <param name="reference">The reference used to generate/update the resource.</param>
+        /// <param name="getResourceFromReference">A function that takes a reference and generates/updates a resource.</param>
+        /// <param name="framesToRemember">The number of frames to remember back.</param>
+        public MemoryBelief(TReference reference, Func<TReference, TObservation> getResourceFromReference, int framesToRemember)
+            : base(reference, getResourceFromReference, () => true)
+        {
+            _memorizedObservations = new(framesToRemember);
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="MemoryBelief{TReference, TResource}"/> class.
+        /// </summary>
+        /// <param name="reference">The reference used to generate/update the resource.</param>
+        /// <param name="getResourceFromReference">A function that takes a reference and generates/updates a resource.</param>
+        /// <param name="framesToRemember">The number of frames to remember back.</param>
+        /// <param name="shouldUpdate">A function that sets a condition on when the resource should be updated.</param>
+        public MemoryBelief(TReference reference, Func<TReference, TObservation> getResourceFromReference, int framesToRemember,
+            Func<bool> shouldUpdate)
+            : base(reference, getResourceFromReference, shouldUpdate)
+        {
+            _memorizedObservations = new(framesToRemember);
+        }
+
+        /// <summary>
+        /// Generates/updates the resource.
+        /// Also stores the previous observation in memory.
+        /// </summary>
+        public override void UpdateBelief()
+        {
+            // We use the implicit conversion to TObservation to store the observation
+            _memorizedObservations.Put(this);
+            base.UpdateBelief();
+        }
+
+        /// <summary>
+        /// Gets the most recently memorized resource.
+        /// </summary>
+        /// <returns> The most recent memory of the resource.</returns>
+        public TObservation GetMostRecentMemory() => _memorizedObservations.GetFirst();
+
+        /// <summary>
+        /// Gets the memorized resource at a specific index.
+        /// A higher index means a memory further back in time.
+        /// If the index is out of bounds, returns the closest element that is in bounds.
+        /// </summary>
+        /// <returns> The memory of the resource at the specified index.</returns>
+        public TObservation GetMemoryAt(int index)
+        {
+            if (index < 0) return GetMostRecentMemory();
+            if (index >= _memorizedObservations.Length)
+                return _memorizedObservations[_memorizedObservations.Length - 1];
+            return _memorizedObservations[index];
+        }
```

Comment:
Actually maybe give the user the option to clamp? On second thought an out-of-range exception seems like the 'safer' approach.
```suggestion
        public TObservation GetMemoryAt(int index, bool clamp = false)
        {
            int lastMemoryIndex = _memorizedObservations.Length - 1;
            if (clamp)
                int index = Math.Min(index, lastMemoryIndex);
            else if (index < 0 || index > lastMemoryIndex)
                throw new IndexOutOfRangeException();
            return _memorizedObservations[index];
        }
```
Also, I feel like you would like to return the _actual_ memory at index i. So when you have room for 5 memories and request the memory at index 1, but there is only a saved memory at index 0, then maybe return the memory at index 0 instead of the one at index 4 (if clamped).

Reply:
TODO: fix in later ticket